feat(dropzone): migrating new S2 tokens, more spaced out dashed lines, more illustrations

aramos-adobe opened the pull request:
## Dropzone S2 Migration

Since dropzone shares a lot of illustrated message's new content tokens (typography and spacing), tons of  unused`--mods` were removed. 

The component reflects the [new design](https://www.figma.com/design/eoZHKJH9a3LJkHYCGt60Vb/S2-token-specs?node-id=10141-2822&m=dev) of the borders that includes dash length and dash gap. 

### SVG Border

To support the intention of the design, an SVG element is used. There modifiable tokens to adjust the length of dashed lines and the gap between them. It's understood it may be too much to consume an additional element so there's a fallback where the SVG element is not necessary - dropzone will use the standard `border` CSS property. 

### New mods

`--mod-drop-zone-content-height`

### Removed mods

`--mod-drop-zone-body-color` 
`--mod-drop-zone-body-font-family` 
`--mod-drop-zone-body-font-style` 
`--mod-drop-zone-body-font-weight` 
`--mod-drop-zone-body-line-height` 
`--mod-drop-zone-content-color` 
`--mod-drop-zone-content-edge-to-text` 
`--mod-drop-zone-content-font-size` 
`--mod-drop-zone-heading-color` 
`--mod-drop-zone-heading-font-family` 
`--mod-drop-zone-heading-font-size` 
`--mod-drop-zone-heading-font-style` 
`--mod-drop-zone-heading-font-weight` 
`--mod-drop-zone-heading-line-height` 
`--mod-drop-zone-heading-to-body` 
`--mod-drop-zone-illustration-to-heading` 
`--mod-drop-zone-illustration-to-title`


## How and where has this been tested?

Please tag yourself on the tests you've marked complete to confirm the tests have been run by someone other than the author.

### Validation steps

  - [x] Filled reflects hover state 
  - [x] Filled & Dragged state reflects `replace`
  - [x] WHCM
  - [ ] S2 Tokens are being used
  - [ ] Normal border dashed style with no SVG - remove or comment out `<svg className="spectrum-DropZone-stroke"> `


### Regression testing

Validate:

1. The documentation pages for at least two other components are still loading, including:

- [x] The pages render correctly, are accessible, and are responsive.

4. If components have been modified, VRTs have been run on this branch:

- [x] VRTs have been run and looked at.
- [ ] Any VRT changes have been accepted (by reviewer and/or PR author), or there are no changes.

## Screenshots

<img width="335" alt="Screenshot 2024-12-04 at 4 30 58 PM" src="https://github.com/user-attachments/assets/61d06ce9-0c5a-4fcf-b085-50288da7a8b8">
<img width="350" alt="Screenshot 2024-12-04 at 4 30 21 PM" src="https://github.com/user-attachments/assets/51de51ff-2587-44a4-8b8a-2bd83450884d">
<img width="349" alt="Screenshot 2024-12-04 at 4 30 16 PM" src="https://github.com/user-attachments/assets/711eddb2-ef06-439f-9c4f-f63df2f044ab">


## To-do list

<!-- Put an "x" to indicate you've done each of the following. Add/remove additional tasks, as needed. -->

- [ ] I have read the [contribution guidelines](/.github/CONTRIBUTING.md).
- [x] I have updated relevant storybook stories and templates.
- [x] I have tested these changes in Windows High Contrast mode.
- [ ] If my change impacts **other components**, I have tested to make sure they don't break.
- [ ] If my change impacts **documentation**, I have updated the documentation accordingly.
- [ ] ✨ This pull request is ready to merge. ✨

rise-erpelding commented on components/dropzone/stories/template.js:
```diff
@@ -17,32 +19,69 @@ export const Template = ({
+		id=${ifDefined(id)}
+		role="region"
+		tabindex="0"
+		style=${[isDragged && isFilled ? `background-image:url(${image});` : null, styleMap(customStyles)].filter(Boolean).join(" ")}
```
I'm seeing this `[object Object]` in the inspector, does something here need to be adjusted?
![image](https://github.com/user-attachments/assets/8afe15e7-ed3b-4921-8165-2e2cd174501f)

rise-erpelding reviewed (changes requested):
Gosh, this one was a lot of work too! Leaving some initial thoughts and some questions about the CSS and design spec.

I think you might have made a couple of pushes after I started reviewing, so pardon if some of the comments made aren't up to date!

aramos-adobe commented on components/dropzone/index.css:
```diff
@@ -154,45 +107,65 @@
+	position: absolute;
+	inset-block-start: -2px;
+	inset-inline-start: -2px;
+	inline-size: 101%;
+	block-size: 101.5%;
```
This was to make the stroke of the svg/rect fit. I think I can bring this bag to 100% now that I figured out how the radius of the svg rectangle works

aramos-adobe commented on components/dropzone/index.css:
@rise-erpelding, adding the border-width to to rootClass-stroke-path makes the stroke-width super fat and it doesn't match the intended design

rise-erpelding reviewed (commented):
Hiya!

The tokens are definitely closer to the spec this time around, thanks! There were a few smaller-ish things I noticed on that front:
* For the text area, I still am not seeing `component-height-300` being used on the filled & dragged--I see it set in the CSS, but it looks it's being overridden by `--mod-drop-zone-content-height`
    ![Screenshot 2025-02-14 at 2 27 39 PM](https://github.com/user-attachments/assets/24e2cf12-5dcf-44ca-9d21-5979d46bee3e)
* Wondering about the illustration color still, I'm not sure the gray color is correct (it's supposed to be `neutral-content-color-default` but also that's really much darker than the spec, so I'm questioning that, am I reading it wrong?). The accent color used for the dragged variant seems maybe right, but does this component need any hover states added?
    ![image](https://github.com/user-attachments/assets/04ca1247-2805-4ec9-811c-1a2a952d84f0)

My bigger concern though is about the svg border, it seems like it could potentially be a little fragile, like if I set a mod to adjust the component width or if I increase the component height with more content, it starts to look stretched:
![Screenshot 2025-02-14 at 2 14 32 PM](https://github.com/user-attachments/assets/975d2fbc-b180-47d7-b0d1-689db0e69127)
![Screenshot 2025-02-14 at 2 10 20 PM](https://github.com/user-attachments/assets/61ec0132-6218-4352-adff-5f9b61d9aadb)

I'm also wondering if we can't use the border-width token or the rx/ry tokens because they're making the border look too thick if there's something that maybe would need adjusting with the svg or rect attributes?

I had asked @jawinn about this and he mentioned that there had been a separate conversation with design about how implementing a border with specific dash lengths and dash gaps was really difficult in CSS and he'd thought they'd opted not to do this instead, so I'm curious if this part is maybe a larger conversation that needs to happen about how to support this design or whether we should support this design--also if React has successfully been able to do it as designed.

Depending on how that goes, we could think about leaving the border out as designed for now (like just having a normal dashed border as it was before) and following up--this PR is in a pretty good place without the border and if that's the main thing holding it up, it feels justified to me to break down the work and do it separately. Let me know what you think!

aramos-adobe commented:
@rise-erpelding,

1. `--mod-drop-zone-content-height` is used to define the height of the filled and dragged state because of the background image. I renamed the other token to `--spectrum-drop-zone-component-height` to specify the use case on the Spectrum-Dropzone-button. 

2. The token we're looking for here is `--spectrum-neutral-visual-color` that equates to `spectrum-gray-500` which appears in Figma. The token name is wrong in Figma example.

3. SVG border: I added `vector-effect: non-scaling-stroke` to the rect. This prevents the borders from scaling and stretching both horizontally and vertically. 

Things about the SVG border that come to mind:
- Feasibility of the dash array mods on the SVG stroke
- Are there possible design inconsistencies on the dash gap and length if there are multiple drop zones on a screen

aramos-adobe commented:
<img width="468" alt="Screenshot 2025-02-19 at 9 48 37 AM" src="https://github.com/user-attachments/assets/f57e5f44-8226-4e42-87b8-e96db34f427a" />

rise-erpelding reviewed (changes requested):
Ok, you've convinced me that maybe the svg might work! Here are my latest notes, including my current thoughts on the svg:

#### Icons files
* I think these found their way in by accident and probably need to be removed

#### Docs Page
This could use a few small updates like:
* Showing the dragged variant
* Showing the filled variant - it looks like this maybe is only meant to be available in medium? That's worth noting on the docs page too.

I wish there was more guidance on the guidelines page that we could borrow, but I didn't see much.

#### Forced colors
* Our action button has disappeared in the dragged filled variant!
![Screenshot 2025-02-19 at 3 50 15 PM](https://github.com/user-attachments/assets/6aa93550-e0b7-42ea-b03f-64c41996339a)

#### SVG Border
![Screenshot 2025-02-19 at 2 33 25 PM](https://github.com/user-attachments/assets/5fe0a3bb-12f4-4de2-a97e-562bc2452a7e)
![Screenshot 2025-02-19 at 2 33 40 PM](https://github.com/user-attachments/assets/f722d5d9-2c36-4e36-8bd8-15885569e5db)
* I think the dashing looks pretty good and the `vector-effect` attribute is pretty cool. I did test this with super edge-case long text and while I think the borders hold up really nicely, you can see some warping in the corners, especially for the variants when the border is solid and not dashed, as shown above.
    * Given that, what would be the pros and cons of using a regular border for the dragged variants and an svg border for the dashed border only? It's an idea, not a suggestion, and maybe that isn't the best course of action.
* Other corner rounding things: the `rx` and `ry` attributes set in the svg's `<rect>` are different from what's being set in the CSS. I think ideally, it's nice to have both, with the attributes being a fallback, but they should match.
* If you apply a visible background color to `.spectrum-DropZone`, or if you apply `overflow: hidden;` to `.spectrum-DropZone` it's looking like the outline border is extending about halfway over the bounds of `.spectrum-DropZone`, I'm wondering if we can change the `inset` to 0 and the `inline-size` and `block-size` to 100%? Would that fix the issue? I don't love that the border sits inside of the drop zone in that case (screenshot below) but that seems preferable to extending over the bounds.
![image](https://github.com/user-attachments/assets/5d456156-c191-4230-b7ed-a03a4c208ec0)
* I'm also wondering about that `viewBox` attribute, it doesn't match the actual size of the svg (seems like it's about half?) and I'm wondering if that's the reason `rx` and `ry` need to be divided by 2? I'm wondering though if doing that would present any additional concerns about how the border/corners looks if drop zone has a different size than expected (if we use a mod, or the drop zone content is really long, etc.).
* Last small small thing - should we provide mods for some of the `stroke-path` styles like `rx`/`ry` and the `stroke-dasharray`?

aramos-adobe commented:
@rise-erpelding ,

- viewBox was distorting the rounded corners 🤯 
- Illustration color is using IM illustration color. [Confirmed by design team](https://adobedesign.slack.com/archives/G01P2PHJMC6/p1740167569715399)

<img width="474" alt="Screenshot 2025-02-24 at 10 58 54 AM" src="https://github.com/user-attachments/assets/7b8799f9-4222-46bb-befc-a73a377c6c62" />
<img width="499" alt="Screenshot 2025-02-21 at 2 55 12 PM" src="https://github.com/user-attachments/assets/00ff9841-b091-4b0e-ae4d-df0bdffd2dd6" />
<img width="476" alt="Screenshot 2025-02-21 at 2 55 00 PM" src="https://github.com/user-attachments/assets/5b3310de-81a6-4c79-a232-6ac1474a1e8e" />

rise-erpelding commented on components/dropzone/index.css:
```diff
@@ -144,55 +84,97 @@
 
-.spectrum-DropZone-content {
-	display: var(--mod-drop-zone-content-display, none);
+.spectrum-DropZone:has(.spectrum-DropZone-stroke) {
+	border: none;
+}
```
![](https://media0.giphy.com/media/v1.Y2lkPTc5MGI3NjExd2ZraGpkbnJmMjVuejFxOHhpdDBvYzVpYnAyNHg1Z2V0OWR5cGhxYSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/5GoVLqeAOo6PK/giphy.gif)

Aw this fallback works so nicely 😍

rise-erpelding commented on components/dropzone/index.css:
```diff
@@ -144,55 +84,97 @@
-		/* Updated values for a nested illustrated message when in a high contrast state */
-		/* stylelint-disable-next-line spectrum-tools/no-unused-custom-properties -- used in illustrated */
-		--highcontrast-illustrated-message-illustration-color: var(--highcontrast-drop-zone-illustration-color);
+		--highcontrast-drop-zone-button-color: Highlight;
```
I did notice that in forced-colors in Chrome the button's looking a little transparent, although I will say it didn't look this transparent when I used AssistivLabs to look at real WHCM:

![Screenshot 2025-02-25 at 2 27 17 PM](https://github.com/user-attachments/assets/80aa9873-29fd-4d76-a48e-b776dfc16ec7)
![Screenshot 2025-02-25 at 2 27 25 PM](https://github.com/user-attachments/assets/b09ee75c-2c69-4de4-9224-a63df74b65b6)

I don't think it's a huge issue and what's here mirrors what's in action button, I think, but you could look at whether ButtonText or ButtonFace might work better?

rise-erpelding commented on components/dropzone/stories/dropzone.stories.js:
```diff
@@ -49,16 +50,25 @@ export default {
+	chromatic: { disableSnapshot: true },
+};
+
+export const Filled = Template.bind({});
```
Is this one only medium or does it come in different sizes? If it's only meant to be one size, would you want to add that in here so it's documented?

rise-erpelding commented on .changeset/big-beds-care.md:
```diff
@@ -0,0 +1,21 @@
+
+### Removed mods
+
+`--mod-drop-zone-body-color` `--mod-drop-zone-body-font-family` `--mod-drop-zone-body-font-style` `--mod-drop-zone-body-font-weight` `--mod-drop-zone-body-line-height` `--mod-drop-zone-content-color` `--mod-drop-zone-content-edge-to-text` `--mod-drop-zone-content-font-size` `--mod-drop-zone-heading-color` `--mod-drop-zone-heading-font-family` `--mod-drop-zone-heading-font-size` `--mod-drop-zone-h ...
```
Would you want to put these each on their own line for easier reading?

rise-erpelding reviewed (approved):
THIS IS LOOKING GREAT! I had a couple of really small callouts but nothing that would block approval.

One more tiny callout about the border in WHCM, I did capture the svg border with a non-contrasting color (but the fallback border doesn't do this). It may not be a huge accessibility concern but I'm putting it out there.
![Screenshot 2025-02-25 at 2 25 41 PM](https://github.com/user-attachments/assets/ba4faef6-7b54-4274-acc6-2c4fd8a2c73d)

marissahuysentruyt commented on components/dropzone/index.css:
```diff
@@ -13,126 +13,66 @@
-		/* @deprecation --mod-drop-zone-border-style--dragged will be removed during the S2 migration; please update your code to --mod-drop-zone-border-style-dragged */
-		--mod-drop-zone-border-style: var(--mod-drop-zone-border-style--dragged, var(--mod-drop-zone-border-style-dragged, solid));
+		--mod-drop-zone-border-style: var(--mod-drop-zone-border-style-dragged, var(--mod-drop-zone-border-style-dragged, solid));
 		--mod-drop-zone-background-color: rgba(var(--spectrum-drop-zone-background-color), var(--mod-drop-zone-background-color-opacity, var(--spectrum-drop-zone-background-color-opacity)));
```
This is just an instance where we probably shouldn't be using the `drop-zone-background-color-opacity` token directly in the style definition (but this isn't something you introduced, either).

marissahuysentruyt reviewed (changes requested):
Nice work on this! This one was pretty hairy and super intwined with illustrated message. Thanks for cleaning it all up! 

I have some questions that I didn't quite know where to put...

- My first question is about using the `mod-drop-zone-width`. If I set that, the inner content is then uncentered. Should we add 
```
display: flex;
align-items: center;
justify-content: center;
```
to `.spectrum-DropZone` to prevent/combat that from happening? I guess I'm thinking that if a user needs to use that mod, I wouldn't want them to _also_ have to go into the CSS to recenter the content. I'd just want them to be able to set the mod, and it just works for them. Does that make sense? Is that even a valid concern?

<img width="724" alt="Screenshot 2025-02-28 at 10 43 05 AM" src="https://github.com/user-attachments/assets/2da481e0-6feb-4607-a5ef-07d19f4d3268" />


- I also wanted to double check the color of the upload SVG visual. You've got it set correctly I think, to neutral-content-color-default (an alias for gray-800) like the Figma specs say. But Figma is _actually_ using neutral-visual-color (which is an alias for gray-500) instead. You may have already checked on this, but which one is design expecting? There's definitely a difference since those 2 grays are quite distinct from each other.

Figma:
<img width="415" alt="Screenshot 2025-02-28 at 9 20 12 AM" src="https://github.com/user-attachments/assets/2493938b-d135-4c2d-af33-a83a6c10f71f" />

Code:
<img width="471" alt="Screenshot 2025-02-28 at 9 20 03 AM" src="https://github.com/user-attachments/assets/c6c62ff2-cc78-493c-b07d-c87ef986eb79" />

 
- Last thing I'm curious about is the button color for `isDragged+isFilled`. Currently, the background of the button in the `isDragged+isFilled` state is set to `accent-visual-color` like the design specs. 👍  However...that token is resolving to a different rgb in the browser when I compare between Chrome and Figma. Is that design token incorrect? It seems like that token (maybe in tandem with the background color opacity?) is causing the button colors to change- is that correct? If I remove the `background` style from `.spectrum-DropZone-button` they resolve to the same rgb, but _not `accent-visual-color`. (it's `--spectrum-accent-background-color-default` 🤦‍♀️)

Lemme know if you wanna pair through any of this! I know I left a lot of ideas and questions for you!

aramos-adobe commented on components/dropzone/index.css:
```diff
@@ -13,126 +13,66 @@
+	--mod-drop-zone-content-height: 280px;
+
+	/* Typography */
+	--mod-illustrated-message-title-font-weight: var(--spectrum-title-serif-font-weight);
```
@marissahuysentruyt This was here before some major changes to `spectrum-two` happened. I can get rid of it now I think.

aramos-adobe commented on components/dropzone/index.css:
```diff
@@ -144,55 +84,106 @@
+	pointer-events: none;
+}
+
+.spectrum-DropZone-stroke-path {
```
@marissahuysentruyt I think that's fine. I'll change it to this

aramos-adobe commented on components/dropzone/stories/template.js:
```diff
@@ -17,32 +19,70 @@ export const Template = ({
 	title,
 	description,
 	label,
-	id = getRandomId("dropzone"),
```
@marissahuysentruyt I'm going to bring it back because it's in the base branch. Every component has it as well. I can't recall why it was removed on my end.

aramos-adobe commented on tokens/custom/global-vars.css:
```diff
@@ -61,8 +61,8 @@
 	/* Gradients that do not change with the color theme, for use in static color backgrounds. */
-	--spectrum-examples-gradient-static-black: linear-gradient(45deg, rgb(255, 241, 246), rgb(238, 245, 255));
-	--spectrum-examples-gradient-static-white: linear-gradient(45deg, rgb(64, 0, 22), rgb(14, 24, 67));
+	--spectrum-examples-gradient-static-black: linear-gradient(45deg, rgb(255 241 246), rgb(238 245 255));
+	--spectrum-examples-gradient-static-white: linear-gradient(45deg, rgb(64 0 22), rgb(14 24 67));
```
@marissahuysentruyt I'm going to bring back the `spectrum-two` version hopefully that should be normal again.

aramos-adobe commented on components/dropzone/stories/template.js:
```diff
@@ -17,32 +19,70 @@ export const Template = ({
-			)}
+			<div class="${rootClass}-actions">
+				${isDragged && isFilled
+					? ActionButton({
+						label: label,
+						customClasses: [`${rootClass}-button`],
+						isSelected: true,
+						isEmphasized: true
+					})
+					: Button({
+						label: label,
+						size: size
```
@marissahuysentruyt The only reason why I have it as action button is because it looks that way visually to me 😂 I'll try this out

aramos-adobe replied:
@marissahuysentruyt The corner radius on button is not the same as the one in the design shown in Figma

aramos-adobe reviewed (commented):
@marissahuysentruyt ,

I addressed the color of the illustration last week in a [thread](https://adobedesign.slack.com/archives/G01P2PHJMC6/p1740167569715399). Design team is updating the Figma to match the token colors

marissahuysentruyt reviewed (approved):
This looks really good! I left some really small comments, but I don't think any of them are blocking, so I'll approve 👍 Sounds like there's some updates that are needed on the Figma side, but the code is all pointing to the right values & tokens.
